docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,303 @@
+  ALTER [COLUMN] column_name
+  [SET DATA] TYPE data_type
+  [COLLATE collation]
+  [USING expression]
+  [DEFERRED | COMPLETE | ABORT]
+```
+
+Column-type changes borrow from the
+[F1 schema-change process](https://www.cockroachlabs.com/blog/how-online-schema-changes-are-possible-in-cockroachdb/)
+to allow a column-type change to occur without taking the affected
+table/column offline.  In general, we assume that a column-type change
+will not be an instantaneous process and could involve a non-trivial
+amount of wall time in order to complete.  While a column-type change is
+in progress, reads from that column will return the original type, while
+any writes must be compatible with both the current and future schema.
+Once the type-change work has been completed, the new type becomes
+effective and the old data can be retired.  We allow only a single
+column-type change to be in progress for any given table at a time.
+
+## Conversion classes
+
+Certain classes of type conversions lend themselves to faster or
+less-complicated approaches and are described herein.  A full table of
+type conversions is described later [TODO] in this RFC.
+
+### Trivial
+
+Trivial conversions are those that involve no conversion of on-disk data
+and which cannot create any conflict between existing data and the
+updated schema, such as converting `STRING -> BYTES` or `STRING(10) ->
+STRING(25)`.  These enable a fast-path conversion by simply updating the
+table descriptor.
+
+### Validated
+
+Validated conversions involve no conversion of on-disk data, but do
+require that all existing data is validated against the proposed schema
+change before completion.  This would include conversions such as
+`STRING(100) -> STRING(10)` or `BYTES -> STRING`.  These have a cost
+somewhere between trivial and general conversions, since on-disk data
+does not need to be rewritten, only revalidated.
+
+### Cast
+
+Cast conversions involve applying a built-in cast behavior  to the
+existing column data requiring the on-disk KV data to be rewritten, such
+as `UUID -> STRING`.
+
+### Assisted
+
+Assisted conversions involve applying a `USING expression` to the
+existing column data, such as `FLOAT -> INT USING ceil(column_name)`.
+
+## Indexed columns
+
+Columns that are part of an index are supported, however columns that
+are part of a primary key are explicitly out of scope for this change.
+
+## Two-phase change
+
+Given that a type-change process may take an arbitrary and unpredictable
+amount of time to complete, we support an optional two-phase approach to
+making the schema change visible to SQL clients.  This allows the user
+to choose exactly when a (breaking) change to the table schema will
+become visible, allowing application downtime to be minimized.
+
+The `ALTER TABLE` command is first run with the `DEFERRED` option, which
+begins the type-conversion in the background and returns an associated
+job id.  At any point in time, but generally after the job has
```

Comment:
What's the use of the job id? We don't return them anywhere else, and I might want to add them all at once instead of here so we can decide exactly when, why, and how we want to return job ids.